style.md

```diff
@@ -1557,7 +1557,237 @@ type Client struct {
 </td></tr>
 </tbody></table>
 
-### Use Field Names to initialize Structs
+As a general rule of thumb, embedding should not make types more difficult to
+use, and should be viewed as an "is-a" declaration: the embedding (outer) type
+can function as the embedded (inner) type, and therefore "is a" form of the
+inner type. A type embedding a `sync.Mutex` declares to users that it "is a"
+mutex; a type embedding a `bytes.Buffer` declares that it "is a" buffer.
+
+Embedded types' methods or fields should never be shadowed by first-class
```

Comment:
Never seems like a strong recommendation, I've often found that embedding + shadowing can be useful as a "decorator".

E.g., let's say you have a `io.WriteCloser`, you might add a decorator that tracks how many bytes `Write` was called on:
```go
type countingWriteCloser struct {
  io.WriteCloser
  
  count int
}

func (w *countingWriteCloser) Write(bs []byte) (int, error) {
  w.count += len(bs)
  return w.WriteCloser.Write(bs)
}
```

In tihs case, it'd be reasonable not to embed as well, but want to get your thoughts on this sort of use

Reply:
Yeah, I think "never" is too strong a word. Some of the common workarounds for embedding where we'd essentially be shadowing are things like

```go
type Foo struct{}
func (f Foo) Bar(...) ... {}

type MultiFoo []Foo

func (mf MultiFoo) Bar(...) ... {
  for _, f := range mf {
    collect(f.Bar())
  }
}
```

etc. I guess a good litmus test would be something like, "should not prevent the underlying method from being called normally" or something to that effect - we're basically trying to argue that it's okay to augment functions by hijacking them, but while not opening the door for "shadowing is universally OK".

I'll take a stab at how to better phrase this, but if you have an idea, let me know!

Reply:
Yeah, I think "never" is too strong a word. Some of the common workarounds for embedding where we'd essentially be shadowing are things like

```go
type Foo struct{}
func (f Foo) Bar(...) ... {}

type MultiFoo []Foo

func (mf MultiFoo) Bar(...) ... {
  for _, f := range mf {
    collect(f.Bar())
  }
}
```

etc. I guess a good litmus test would be something like, "should not prevent the underlying method from being called normally" or something to that effect - we're basically trying to argue that it's okay to augment functions by hijacking them, but while not opening the door for "shadowing is universally OK".

I'll take a stab at how to better phrase this, but if you have an idea, let me know!